File: src/apps/daemon/internal/networking/dns/packed_control_block.go
// Package dns provides networking, tunneling, and scheduling primitives.
// Conforms strictly to §8 structural cleanroom rules.

package dns

const (
	// PackedControlBlockSize is the fixed size of each control block:
	// Type(1) + StreamID(2) + SeqNum(2) + FragID(1) + Total(1) = 7 Bytes
	PackedControlBlockSize = 7
)

// ControlBlock describes a packed ACK, NACK, SYN_ACK, or RST_ACK block.
type ControlBlock struct {
	PacketType     uint8  `json:"packet_type"`
	StreamID       uint16 `json:"stream_id"`
	SequenceNum    uint16 `json:"sequence_num"`
	FragmentID     uint8  `json:"fragment_id"`
	TotalFragments uint8  `json:"total_fragments"`
}

// AppendPackedControlBlock serializes a block and appends the 7 bytes to dst.
func AppendPackedControlBlock(dst []byte, packetType uint8, streamID uint16, seqNum uint16, fragID uint8, total uint8) []byte {
	return append(dst,
		packetType,
		uint8(streamID>>8), uint8(streamID),
		uint8(seqNum>>8), uint8(seqNum),
		fragID,
		total,
	)
}

// ForEachPackedControlBlock parses contiguous 7-byte blocks in payload.
func ForEachPackedControlBlock(payload []byte, yield func(block ControlBlock) bool) {
	if len(payload) < PackedControlBlockSize || yield == nil {
		return
	}

	for offset := 0; offset+PackedControlBlockSize <= len(payload); offset += PackedControlBlockSize {
		b := ControlBlock{
			PacketType:     payload[offset],
			StreamID:       uint16(payload[offset+1])<<8 | uint16(payload[offset+2]),
			SequenceNum:    uint16(payload[offset+3])<<8 | uint16(payload[offset+4]),
			FragmentID:     payload[offset+5],
			TotalFragments: payload[offset+6],
		}
		if !yield(b) {
			break
		}
	}
}

// ParsePackedControlBlocks extracts all control blocks from a packed payload.
func ParsePackedControlBlocks(payload []byte) []ControlBlock {
	var blocks []ControlBlock
	ForEachPackedControlBlock(payload, func(b ControlBlock) bool {
		blocks = append(blocks, b)
		return true
	})
	return blocks
}

// PackControlBlocks serializes multiple control blocks into a contiguous byte slice.
func PackControlBlocks(blocks []ControlBlock) []byte {
	dst := make([]byte, 0, len(blocks)*PackedControlBlockSize)
	for _, b := range blocks {
		dst = AppendPackedControlBlock(dst, b.PacketType, b.StreamID, b.SequenceNum, b.FragmentID, b.TotalFragments)
	}
	return dst
}

File: src/packages/lumicore/src/transport/packed_control_block.rs
//! # Packed Control Block Framing
//!
//! Multiplexes control signals (ACK, NACK, SYN_ACK, RST_ACK, etc.) into compact 7-byte blocks:
//! `Type(1) + StreamID(2) + SeqNum(2) + FragID(1) + Total(1) = 7 Bytes`.
//! Reduces DNS query packet count by packing dozens of control updates per datagram.
//! Conforms to §8 structural cleanroom rules.

pub const PACKED_CONTROL_BLOCK_SIZE: usize = 7;

/// A parsed control block extracted from a packed control datagram.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ControlBlock {
    pub packet_type: u8,
    pub stream_id: u16,
    pub sequence_num: u16,
    pub fragment_id: u8,
    pub total_fragments: u8,
}

impl ControlBlock {
    pub fn new(packet_type: u8, stream_id: u16, sequence_num: u16, fragment_id: u8, total_fragments: u8) -> Self {
        Self {
            packet_type,
            stream_id,
            sequence_num,
            fragment_id,
            total_fragments,
        }
    }

    /// Serializes block into 7-byte big-endian representation.
    pub fn encode(&self) -> [u8; PACKED_CONTROL_BLOCK_SIZE] {
        [
            self.packet_type,
            (self.stream_id >> 8) as u8,
            self.stream_id as u8,
            (self.sequence_num >> 8) as u8,
            self.sequence_num as u8,
            self.fragment_id,
            self.total_fragments,
        ]
    }

    /// Appends the 7-byte serialized block directly to a destination byte buffer.
    pub fn append_to(&self, dst: &mut Vec<u8>) {
        dst.extend_from_slice(&self.encode());
    }

    /// Deserializes a single 7-byte slice.
    pub fn decode(src: &[u8]) -> Option<Self> {
        if src.len() < PACKED_CONTROL_BLOCK_SIZE {
            return None;
        }
        let packet_type = src[0];
        let stream_id = ((src[1] as u16) << 8) | (src[2] as u16);
        let sequence_num = ((src[3] as u16) << 8) | (src[4] as u16);
        let fragment_id = src[5];
        let total_fragments = src[6];

        Some(Self {
            packet_type,
            stream_id,
            sequence_num,
            fragment_id,
            total_fragments,
        })
    }
}

/// Iterates through contiguous 7-byte blocks in a packed payload.
pub fn parse_packed_control_blocks(payload: &[u8]) -> Vec<ControlBlock> {
    let mut blocks = Vec::new();
    let count = payload.len() / PACKED_CONTROL_BLOCK_SIZE;
    for i in 0..count {
        let offset = i * PACKED_CONTROL_BLOCK_SIZE;
        if let Some(block) = ControlBlock::decode(&payload[offset..offset + PACKED_CONTROL_BLOCK_SIZE]) {
            blocks.push(block);
        }
    }
    blocks
}

/// Packs multiple control blocks into a single byte slice.
pub fn pack_control_blocks(blocks: &[ControlBlock]) -> Vec<u8> {
    let mut dst = Vec::with_capacity(blocks.len() * PACKED_CONTROL_BLOCK_SIZE);
    for b in blocks {
        b.append_to(&mut dst);
    }
    dst
}
